### ISSUE TITLE
Lab 8 Test case 8_4 issues

### ISSUE THREAD

**Issue by Aaronpeh022** [nus-cs2030/2223-s1#327]
# Summary
Able to clear test case 6_1 to 8_3 but the output for test case 8_4 is wrong

## Description:
My output displays customer cutting queue even though it is not their turn yet. Does anyone encounter similar problem?

## Code Snippets

Output:
<img width="415" alt="Screenshot 2022-11-28 204643" src="https://user-images.githubusercontent.com/87894275/204281445-31d472e4-65c6-4071-9f48-32df3761e65a.png">

**Comment by cherylchan02** [comment]
It seems as if the problem may be in the way the customers are being queued/removed from queue. Looks like both customer 14 and 15 are queuing at your self-check counter but the order that they are removed from the queue to be served is wrong. If you have an actual queue within your self-check it might be good to print it out to see if the customers are being added/removed from the queue correctly.

**Comment by OngMinXian** [comment]
I encountered this problem too. I noticed that when I put 14 and 15 into my PQ, 15 had a lower event time even though 14 was supposed to be at the front of the queue. I added a check whenever I pop customers from the self check queue if they are indeed at the front of the queue, or not I will requeue them. In this case, I would requeue 15 to a later time.
Hopefully this was helpful

**Comment by izascia256** [comment]
Might want to check how you are assigning the self checkout counters during the Wait Event. You might have assigned a self checkout counter during the Wait Event, which causes the queue to no longer be collective.

**Comment by BrianNeo00** [comment]
I had a similar issue as well, though it was because of how i handled my queue in the self-checkout. I made it such that events were assigned the self-checkout server, which then determined the best counter to go to, and was updated when the event is executed. In my case there was a bug with the the queue handling as the reference self-checkout in my simulator was not updating properly. If your implementation is something like mine you may want to look at that

**Comment by shecheeyee** [comment]
I had this exact same issue. What happened was that while the previous customer was in the self-checkout servers (all are full), I preassigned which counter and service time customer 14 will have based on the earliest available time among the self-checkout servers. When customer 15 arrived (keeping in mind that all self-checkout servers are still busy), the same counter and service time is given to customer 15. This somehow messed up the comparison and resulted in customer 15 going first. Customer 14 is updated to a later time.

What I did to resolve this was to implement a check and update the timing for all the customers (or rather events) in the queue so that such an error won't appear. Kind of similar to updating the scheduled `Serve` events in lab 5/6 each time there is a `Done` event.
